request_reader.go

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package main
+
+import (
+	"encoding/json"
+	"io"
+
+	gyaml "github.com/ghodss/yaml"
+	"gopkg.in/yaml.v2"
+)
+
+type requestReader interface {
+	next() ([]byte, error)
+}
+
+// jsonRequestReader parses multiple JSON objects from given reader
+// JSON objects can be delimited by space or newline
+type jsonRequestReader struct{ *json.Decoder }
+
+func (r *jsonRequestReader) next() ([]byte, error) {
+	if !r.More() {
+		return nil, io.EOF
+	}
+	var v json.RawMessage
+	err := r.Decode(&v)
+	if err != nil {
+		return nil, err
+	}
+	return []byte(v), nil
+}
+
+// yamlRequestReader parses multiple YAML objects from given reader
+// YAML objects must be delimited by `---`
+type yamlRequestReader struct{ *yaml.Decoder }
+
+// next decodes the request from the reader as YAML or JSON and returns
+// JSON encoded byte
+func (r *yamlRequestReader) next() ([]byte, error) {
+	var v interface{}
+	err := r.Decode(&v)
+	if err != nil {
+		return nil, err
+	}
+	bytes, err := yaml.Marshal(v)
+	if err != nil {
+		return nil, err
+	}
+	return gyaml.YAMLToJSON(bytes)
+}
+
+// isJSONInput returns true if input reader can be parsed into JSON correctly
+func isJSONInput(reader io.Reader) bool {
+	var v json.RawMessage
+	jsonDecoder := json.NewDecoder(reader)
+	err := jsonDecoder.Decode(&v)
+	return err == nil
+}
+
+// newRequestReader detects the input encoding type, returns either
+// json or yaml request decoder
+// note: wraps reader under buffer to seek to the beginning for readers
+// which do not support seeking such as io.STDIN
+func newRequestReader(reader io.Reader) requestReader {
+	bufReader := newBufferReader(reader)
```

Comment:
This method of determining whether the input is JSON is a bit expensive and requires maintaining a custom bufferReader (which has unbounded memory growth for streaming inputs).

We can perhaps simplify this all significantly with a simpler heuristic and with use of [bufio.Reader](https://golang.org/pkg/bufio/#Reader), which supports [`Peek`](https://golang.org/pkg/bufio/#Reader.Peek) and [`UnreadByte`](https://golang.org/pkg/bufio/#Reader.UnreadByte).

The heuristic can be: does the input start with `{`? It's not perfect but it's good enough. If we want something smarter later, we can build something that accepts both, JSON and YAML on the same stream. But for now, we can do:

```
func isJSONInput(r *bufio.Reader) (bool, error) {
  b, err := r.ReadByte()
  ...
  isJSON := b == '{'
  err := r.UnreadByte()
  ...
```

That way we're not double decoding, nor are do we have unbounded memory growth on a custom buffered reader.

Reply:
+1 unbounded memory growth is concerning, this would break for many YAML inputs like `{a:b}` as decoder will consider it as JSON. Since the buffer capture is required only on the initial object parsing, I could parse it using the custom reader and post encoding detection it could stop caching the data and work only as pass through cache, this should avoid unbounded memory growth. Thoughts?

Reply:
I'm concerned that short of something super custom and complicated, any version that doesn't attempt full decoding is not going to be perfect because all JSON is valid YAML.

I would maybe opt for a simple-ish heuristic at the cost of imperfect detection for now.

Reply:
Makes sense, will go ahead with simple decoder for now :-)